clang-tools-extra/clang-tidy/readability/UseBuiltinLiteralsCheck.cpp

```diff
@@ -0,0 +1,171 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "UseBuiltinLiteralsCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Lex/Lexer.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::readability {
+
+namespace {
+
+static StringRef getRawStringRef(const SourceRange &Range,
+                                 const SourceManager &Sources,
+                                 const LangOptions &LangOpts) {
+  CharSourceRange TextRange = Lexer::getAsCharRange(Range, Sources, LangOpts);
+  return Lexer::getSourceText(TextRange, Sources, LangOpts);
+}
+
+typedef bool (*RuleOnStd)(const LangStandard &LS);
+
+struct Replacement {
+  Replacement(const StringRef &_Seq, const RuleOnStd _Std = NULL)
+      : Seq(_Seq), Std(_Std) {}
+  bool operator()(const LangOptions &LO) const {
+    return Std ? Std(LangStandard::getLangStandardForKind(LO.LangStd)) : true;
+  }
+  StringRef Seq;
+  RuleOnStd Std;
+};
+
+llvm::Regex CharRegex("^(u8|u|U|L)?");
+llvm::StringMap<Replacement> CharPrefix({
+    {"char", {""}},
+    {"char8_t", {"u8"}},
+    {"char16_t", {"u"}},
+    {"char32_t", {"U"}},
+    {"wchar_t", {"L"}},
+});
+
+llvm::Regex
+    IntRegex("(([uU]?[lL]{0,2})|([lL]{0,2}[uU]?)|([uU]?[zZ]?)|([zZ]?[uU]?))?$");
+llvm::StringMap<Replacement> IntSuffix({
+    {"int", {""}},
+    {"unsigned int", {"u"}},
+    {"long", {"L"}},
+    {"unsigned long", {"uL"}},
+    {"long long", {"LL"}},
+    {"unsigned long long", {"uLL"}},
+    {"size_t", {"uz", [](const auto &LS) { return LS.isCPlusPlus23(); }}},
+    {"std::size_t", {"uz", [](const auto &LS) { return LS.isCPlusPlus23(); }}},
+});
+
+llvm::Regex FloatRegex(
+    "([fF]|[lL]|([fF]16)|([fF]32)|([fF]64)|([fF]128)|((bf|BF)16))?$");
+llvm::StringMap<Replacement> FloatSuffix({
+    {"double", {""}},
+    {"float", {"f"}},
+    {"long double", {"L"}},
+    {"std::float16_t", {"f16"}},
+    {"std::float32_t", {"f32"}},
+    {"std::float64_t", {"f64"}},
+    {"std::float128_t", {"f128"}},
+    {"std::bfloat16_t", {"bf16"}},
+    {"float16_t", {"f16"}},
+    {"float32_t", {"f32"}},
+    {"float64_t", {"f64"}},
+    {"float128_t", {"f128"}},
+    {"bfloat16_t", {"bf16"}},
+});
+
+} // namespace
+
+void UseBuiltinLiteralsCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      traverse(TK_IgnoreUnlessSpelledInSource,
+               explicitCastExpr((has(ignoringParenImpCasts(
+                                    anyOf(characterLiteral().bind("char"),
+                                          integerLiteral().bind("int"),
+                                          floatLiteral().bind("float"))))))
+                   .bind("expr")),
+      this);
+  Finder->addMatcher(
+      traverse(TK_IgnoreUnlessSpelledInSource,
+               explicitCastExpr(has(initListExpr(has(ignoringParenImpCasts(
```

Comment:
Just a question because I see a lot of the innards of the matcher are duplicated here.

First, the inner matcher should be factored out to only be defined once. The outer `expr()` is needed because `anyOf()` is variadic.

```cpp
static const auto Literal = expr(anyOf(characterLiteral().bind("char"), ...));
```

Following that won't `traverse(..., explicitCastExpr(has(anyOf(implicitCastExpr(Literal), expr(Literal)))))` work, so the almost similar matcher does not need to be defined twice?

Reply:
Yes, and thank you. Now it's factored out into a separate matcher, and binding to the `expr()` before `anyOf()` also let me factor out the reading of the literal's textual representation in the check's logic.